presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java

```diff
@@ -0,0 +1,1970 @@
+                            Partition partition = action.getData().getPartition();
+                            Path targetPath = new Path(partition.getStorage().getLocation());
+                            Path currentPath = action.getData().getCurrentLocation();
+                            cleanUpTasks.add(new DirectoryCleanUpTask(DirectoryTaskKind.MOVE_BACKWARD, user, targetPath, false));
+                            if (!targetPath.equals(currentPath)) {
+                                asyncRename(hdfsEnvironment, renameExecutor, fileRenameCancelled, fileRenameFutures, user, currentPath, targetPath, action.getData().getFileNames());
+                            }
+                            break;
+                        }
+                        default:
+                            throw new IllegalStateException("Unknown action type");
+                    }
+                }
+            }
+
+            // Wait for all renames submitted for "INSERT_EXISTING" action to finish
+            for (CompletableFuture<?> fileRenameFuture : fileRenameFutures) {
+                MoreFutures.getFutureValue(fileRenameFuture, PrestoException.class);
+            }
+
+            // At this point, all file system operations, whether asynchronous issued or not, must complete successfully.
+            // We are moving on to metastore operations now.
+
+            for (CreateTableOperation addTableOperation : addTableOperations) {
+                addTableOperation.run(delegate);
+            }
+
+            for (AlterPartitionOperation alterPartitionOperation : alterPartitionOperations) {
+                alterPartitionOperation.run(delegate);
+            }
+
+            for (PartitionAdder partitionAdder : partitionAdders) {
+                partitionAdder.execute();
+            }
+
+            // After this line, operations are no longer reversible.
+            // The operations that haven't been carried out yet are
+            // * dropping table/partition, and
+            // * cleanups.
+            // The next section will deal with "dropping table/partition". Commit may still fail in
+            // this section. Even if commit fails, cleanups, instead of rollbacks, will be executed.
+            moveForwardInFinally = true;
+
+            List<String> failedIrreversibleOperationDescriptions = new ArrayList<>();
+            Throwable lastThrowableFromIrreversibleOperations = null;
+            for (IrreversibleMetastoreOperation irreversibleMetastoreOperation : metastoreDeleteOperations) {
+                try {
+                    irreversibleMetastoreOperation.run();
+                }
+                catch (Throwable t) {
+                    failedIrreversibleOperationDescriptions.add(irreversibleMetastoreOperation.getDescription());
+                    lastThrowableFromIrreversibleOperations = t;
+                }
+            }
+            if (lastThrowableFromIrreversibleOperations != null) {
```

Comment:
Instead of making `lastThrowableFromIrreversibleOperations` the cause, I would add add `failedIrreversibleOperationDescriptions` as suppressed exceptions.  Then you can remove the `failedIrreversibleOperationDescriptions` variable